tests/unit_tests/test_internal_api.py

```diff
@@ -72,3 +76,283 @@ def test_download_write_file_fetches_iff_file_checksum_mismatched(
             assert response is not None
         else:
             assert response is None
+
+
+@pytest.fixture
+def some_file(tmp_path: Path):
+    p = tmp_path / "some_file.txt"
+    p.write_text("some text")
+    return p
+
+
+MockResponseOrException = Union[Exception, Tuple[int, Mapping[int, int], str]]
+
+
+class TestUploadFile:
+    class TestSimple:
+        def test_adds_headers_to_request(
+            self, mock_responses: responses.RequestsMock, some_file: Path
+        ):
+            response_callback = Mock(return_value=(200, {}, "success!"))
+            mock_responses.add_callback(
+                "PUT", "http://example.com/upload-dst", response_callback
+            )
+            internal.InternalApi().upload_file(
+                "http://example.com/upload-dst",
+                some_file.open("rb"),
+                extra_headers={"X-Test": "test"},
+            )
+            assert response_callback.call_args[0][0].headers["X-Test"] == "test"
+
+        def test_returns_response_on_success(
+            self, mock_responses: responses.RequestsMock, some_file: Path
+        ):
+            mock_responses.add(
+                "PUT", "http://example.com/upload-dst", status=200, body="success!"
+            )
+            resp = internal.InternalApi().upload_file(
+                "http://example.com/upload-dst", some_file.open("rb")
+            )
+            assert resp.content == b"success!"
+
+        @pytest.mark.parametrize(
+            "response,expected_errtype",
+            [
+                ((400, {}, ""), requests.exceptions.HTTPError),
+                ((500, {}, ""), retry.TransientError),
+                ((502, {}, ""), retry.TransientError),
+                (requests.exceptions.ConnectionError(), retry.TransientError),
+                (requests.exceptions.Timeout(), retry.TransientError),
+                (RuntimeError("oh no"), RuntimeError),
+            ],
+        )
+        def test_returns_transienterror_on_transient_issues(
+            self,
+            mock_responses: responses.RequestsMock,
+            some_file: Path,
+            response: MockResponseOrException,
+            expected_errtype: Type[Exception],
+        ):
+            mock_responses.add_callback(
+                "PUT",
+                "http://example.com/upload-dst",
+                lambda _: response,
```

Comment:
probably ok, but just verify the late binding of the lambda function here

Reply:
Ah, sure, legit. I'll replace it with `Mock(return_value=response)`.